central/sensor/service/service_impl.go

```diff
@@ -178,6 +187,15 @@ func (s *serviceImpl) Communicate(server central.SensorService_CommunicateServer
 		return nil
 	}
 
+	if cluster.GetHealthStatus().GetLastContact() == nil {
+		// Sensor has initially connected with a real service certificate, not just with a CRS.
+		// This also updates the revocation state of the CRS used for this cluster, if needed.
+		if err := s.clusterInitStore.RecordCompletedRegistration(clusterDSSAC, cluster.GetInitBundleId()); err != nil {
```

Comment:
(Similar to the problem with lack of transactionality when bumping registrations initiated).

AFAICT if something goes wrong here between this line and the place where `lastContact` is set, we will double-book the completed registrations on next connection, effectively preventing some clusters from completing the registration.

Reply:
I think I got lost... maybe we don't need to use lastContact at all if we're treating initBundleId as the guard? 🤔 
It would be good to look at a state transition diagram...